sw/device/tests/rv_timer_systick_test.c

```diff
@@ -0,0 +1,93 @@
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_rv_timer_t timer;
+
+enum {
+  kHart = kTopEarlgreyPlicTargetIbex0,
+  kComparator = 0,
+  kReferenceTimeMicros = 1000 * 1000,  // 1 second
+};
+
+OTTF_DEFINE_TEST_CONFIG();
+
+status_t set_tick(uint32_t tick_hz) {
+  dif_rv_timer_tick_params_t tick_params;
+  TRY(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz, tick_hz,
+                                           &tick_params));
+  TRY(dif_rv_timer_set_tick_params(&timer, kHart, tick_params));
+  return OK_STATUS();
+}
+
+status_t test_tick(uint32_t tick_hz) {
+  LOG_INFO("%s: tick_hz = %u", __func__, tick_hz);
+  uint64_t counter = 0;
+
+  TRY(set_tick(tick_hz));
+
+  TRY(dif_rv_timer_counter_write(&timer, kHart, 0));
+  TRY(dif_rv_timer_counter_read(&timer, kHart, &counter));
+  TRY_CHECK(counter == 0, "Failed to write the counter");
+
+  TRY(dif_rv_timer_counter_set_enabled(&timer, kHart, kDifToggleEnabled));
+  busy_spin_micros(kReferenceTimeMicros);
+  TRY(dif_rv_timer_counter_set_enabled(&timer, kHart, kDifToggleDisabled));
+
+  TRY(dif_rv_timer_counter_read(&timer, kHart, &counter));
+  const uint64_t elapsed_micros = udiv64_slow(counter, tick_hz, NULL) * 1000000;
+
+  // Verify that `n * T ~= 1 second` within 3% of tolerance.
+  TRY_CHECK((elapsed_micros > (uint64_t)(kReferenceTimeMicros * 0.97)) &&
+                (elapsed_micros < (uint64_t)(kReferenceTimeMicros * 1.03)),
+            "Unexpected elapsed time, expected: %u, got: %u",
+            (uint32_t)kReferenceTimeMicros, (uint32_t)elapsed_micros);
+
+  return OK_STATUS();
+}
+
+/**
+ * Verify that the timer can be configured to generate a system tick.
+ * - Configure the timer to generate a tick of `T` microseconds long.
+ * - Enable the timer.
+ * - Start a busy loop of 1 second based on the `mcycleh` CSR.
+ * - Read the number of ticks `n`.
+ * - Verify that `n * T ~= 1 second` within 3% of tolerance.
+ * - Repeat for T in:
+ *     [1, 5, 10, 25, 50, 100, 125, 250, 500]
+ */
+bool test_main(void) {
+  CHECK_DIF_OK(dif_rv_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR), &timer));
+  CHECK_DIF_OK(dif_rv_timer_reset(&timer));
+
+  status_t result = OK_STATUS();
+
+  const uint32_t kTickHz[] = {
+      1 * 1000 * 1000,  // 1MHz - 1us.
+      200 * 1000,       // 200kHz - 5us.
+      100 * 1000,       // 100kHz - 10us.
+      40 * 1000,        // 40kHz - 25us.
+      20 * 1000,        // 20kHz - 50us.
+      10 * 1000,        // 10kHz - 100us.
+      8 * 1000,         // 8kHz - 125us.
+      4 * 1000,         // 4kHz - 250us.
+      2 * 1000,         // 2kHz - 500us.
+  };
+
+  for (size_t i = 0; i < ARRAYSIZE(kTickHz); ++i) {
```

Comment:
Imagine the values i >= ARRAYSIZE(..) - 2 take too long to simulate. Include lib/arch/device.h, and do something like:

  size_t end_tick = kDeviceType == kDeviceSimDV ? ARRAYSIZE(..) - 2 : ARRAYSIZE(..);
  for (size_t i = 0; i < end_tick; ++i) {
    ....
  }